browser_use/browser/session.py

```diff
@@ -343,24 +350,21 @@ async def setup_browser_via_passed_objects(self) -> None:
 		"""Override to customize the set up of the connection to an existing browser"""
 
 		# 1. check for a passed Page object, if present, it always takes priority, set browser_context = page.context
-		self.browser_context = (self.agent_current_page and self.agent_current_page.context) or self.browser_context or None
+		if self.agent_current_page:
+			self.browser_context = self.agent_current_page.context
 
 		# 2. Check if the current browser connection is valid, if not clear the invalid objects
-		if self.browser_context:
-			try:
-				# Try to access a property that would fail if the context is closed
-				_ = self.browser_context.pages
-				# Additional check: verify the browser is still connected
-				if self.browser_context.browser and not self.browser_context.browser.is_connected():
-					self.browser_context = None
-			except Exception:
-				# Context is closed, clear it
-				self.browser_context = None
-
-		# 3. if we have a browser object but it's disconnected, clear it and the context because we cant use either
-		if self.browser and not self.browser.is_connected():
-			if self.browser_context and (self.browser_context.browser is self.browser):
-				self.browser_context = None
+		try:
+			assert self.browser is not None, 'browser is not set'
+			assert self.browser.is_connected(), 'browser is not connected'
+			assert self.browser_context is not None, 'browser_context is not set'
+			assert self.browser_context.browser == self.browser, 'browser_context.browser != browser'
+			# Try to access a property that would fail if the context is closed
+			_ = self.browser_context.pages
+			# if code reaches here, the browser connection is valid
+		except Exception:
+			# the browser connection is invalid, clear browser_context and browser
+			self.browser_context = None
```

Comment:
`self.browser` is allowed to be `None` for a valid & connected `BrowserSession`. We only need a `browser_context` actually, and the only way to check a `browser_context` is connected is to try to use it.

Playwright does not actually provide a `Browser` object when `launch_persistent_context()` is used (which causes all these nightmares for us, terrible API design on their part imo)

<img width="1324" alt="image" src="https://github.com/user-attachments/assets/11f9e4ea-ea71-42c9-9f7b-8c2e65bb1b0a" />